### ISSUE TITLE
Cortex content-addressed node store: addresses, execution records, memoization

### ISSUE THREAD

**Issue by juliuskoskela** [Digimuoto/cortex#25]
## Goal

Replace authored-name NodeId with a content-addressed identity scheme for Cortex nodes. Enables pure-node memoization, cross-run lineage, and edit-resilient replay without coordinating allocated IDs.

## Addressing

```haskell
newtype NodeAddress = NodeAddress ContentHash
 -- hash(canonicalEncode spec, sortByPortName predecessors)

data ExecutionId
 = ExecutionId NodeAddress RunId Attempt
```

Two identity scopes:

* **NodeAddress** — graph-position identity, stable across runs. Same spec + same predecessors anywhere in any wire = same address.
* **ExecutionId** — per-run invocation identity. Pure nodes collapse this to NodeAddress (output is deterministic); IO nodes keep the full tuple.

## Hash inputs

Canonical encoding of spec must include:

* executor tag (act / await / decide / composite)
* executor config:
 * for actPure / actIO: prompt template content-hash, model + weights hash, seed strategy, sampling config, tool set sorted by name
 * for await: signal key
 * for decide: decision function hash (via Typeable or hand-registered)
* output contract id (via WireContract instance)
* port declarations (names + contract ids)

Predecessors are a Map InputPort NodeAddress. Serialize sorted by port name so join-order in source does not affect the hash.

## Store semantics

| Node kind | Address maps to | Second invocation with same inputs |
| -- | -- | -- |
| NodePure | Cached output (single entry) | Cache hit, skip execution |
| NodeIO | List of ExecutionRecords | New execution record appended |

```haskell
storeLookup :: NodeAddress -> StoreResult
storeWrite :: ExecutionRecord -> IO
```

The purity flag lives on the Node constructor; the store reads it and branches. Same address scheme, different storage semantics.

## Cache-on-success rule

**The cache writes iff the outcome is** `Completed (Right o)` (see #24 Outcome).

Skipped:

* `Completed (Left e)` — a node's error path may be non-deterministic (model refused, tool timed out provider-specifically, prompt injection detected); caching the error would be unsafe even for otherwise-pure nodes. Domain errors represent a decision about behavior, not a computed value that is identity-stable.
* `Halted _ _` — the execution did not finish computing. Gas exhaustion carries no claim about what the completed output would have been.

Single rule, two skipped cases. Triple-boundary: **store populates only on successful value output**. This applies uniformly to pure and IO nodes. The purity flag determines cache eligibility; the outcome shape determines whether a specific invocation writes.

## Seed strategies for actPure

```haskell
data SeedStrategy
 = FixedSeed Seed -- pinned; ablations, classifier uses
 | DerivedFromInput -- seed = hash(input); memoizable across runs (recommended default)
 | DerivedFromRunId -- seed = hash(runId, nodeAddress); variety across runs, replayable within a run
```

DerivedFromInput is the choice that makes the content-addressed store maximally effective for topology experiments — entire parallel analyst fan-outs cache against the evidence-batch hash.

## Identity split at system boundary

Content-hashed internally, UUID at the boundary:

* **Hash**: node addresses, execution records, derivation output, persisted artifact content, prompt template versions
* **UUID**: RunId, workspace item ids, session ids, user tokens, URL-shaped identifiers

RunId is a UUID precisely because a run is an allocation event, not a computation; it mints fresh identity and seeds DerivedFromRunId.

## Gas is not part of the address

Two executions of the same node with different gas envelopes are the same computation, same cache address. Gas is a runtime allocation concern, not an identity concern. See #26 for the gas model.

## Model-weight ripple

lmHash in actPure's closure → shipping a new model version invalidates every cached pure-LLM result keyed on that model. This is correct behavior. Design the store with model-upgrade GC events in mind; a model revision should be a deliberate action, not incidental.

## Prompt template hashing

Prompt template id alone is insufficient — include a content hash of the template. Otherwise stale cache hits on template edits. Templates are versioned by content, not authored id.

## Floating-point determinism caveat

"Pure LLM" is only as strong as the inference engine's determinism. GPU kernel reductions, batch composition, mixed-precision quantization can shift logits. For cache-keying purposes, "deterministic enough to key" is the practical bar — bit-exactness is harder. lmHash should include (weights digest, kernel version, hardware class) to be safe.

## Precedent

* Nix: input-addressed derivation + content-addressed build output + fixed-output derivations for impure fetches. Our NodePure / NodeIO split maps cleanly.
* Git: content-addressed blobs/trees/commits + ref-name pointers for allocation identity. Our RunId plays the role of a ref; node addresses are the blob/tree layer.

## What this kills

* Cortex.Pulse.Plan's StageActionId / StageTemplateId newtypes become derivable, not authored.
* Authoring-name-based "re-run the failed run" machinery: becomes trivially correct via address matching.
* Template-registry indirection beyond the spec hash: same spec anywhere = same address.

## Parent

* #23 (epic)

## Related

* #24 (Node algebra) — node spec structure is what gets canonically encoded
* #13 (capability registry) — capability spec contributes to the node-spec hash
* #26 (gas) — gas excluded from the address hash
